src/lib/multi_mappings.rs

```diff
@@ -0,0 +1,305 @@
+    pub fn push(&self, src: NodeId<SrcNodeId>, dst: NodeId<DstNodeId>) {
+        if !self.src.borrow().contains_key(&src) {
+            self.src.borrow_mut().insert(src, HashSet::new());
+        }
+        self.src.borrow_mut().get_mut(&src).unwrap().insert(dst);
+        if !self.dst.borrow().contains_key(&dst) {
+            self.dst.borrow_mut().insert(dst, HashSet::new());
+        }
+        self.dst.borrow_mut().get_mut(&dst).unwrap().insert(src);
+        info!("Mapping: {:?} {:} -> {:?} {:}",
+              self.src_arena.borrow()[src].ty,
+              self.src_arena.borrow()[src].label,
+              self.dst_arena.borrow()[dst].ty,
+              self.dst_arena.borrow()[dst].label);
+    }
+
+    /// Remove mapping from store.
+    pub fn remove(&self, src: NodeId<SrcNodeId>, dst: NodeId<DstNodeId>) {
+        self.src.borrow_mut().get_mut(&src).unwrap().remove(&dst);
+        if self.src.borrow().get(&src).unwrap().is_empty() {
+            self.src.borrow_mut().remove(&src);
+        }
+        self.dst.borrow_mut().get_mut(&dst).unwrap().remove(&src);
+        if self.dst.borrow().get(&dst).unwrap().is_empty() {
+            self.dst.borrow_mut().remove(&dst);
+        }
+    }
+
+    /// `true` if the store has a mapping from `src` to other nodes.
+    pub fn contains_src(&self, src: NodeId<SrcNodeId>) -> bool {
+        self.src.borrow().contains_key(&src)
+    }
+
+    /// `true` if the store has a mapping from a node to `dst`.
+    pub fn contains_dst(&self, dst: NodeId<DstNodeId>) -> bool {
+        self.dst.borrow().contains_key(&dst)
+    }
+
+    /// `true` if the store has a mapping from `src` to `dst`.
+    pub fn contains_mapping(&self, src: NodeId<SrcNodeId>, dst: NodeId<DstNodeId>) -> bool {
+        self.src.borrow().contains_key(&src) && self.src.borrow().get(&src).unwrap().contains(&dst)
+    }
+
+    /// Return a set of all nodes mapped to `dst`.
+    pub fn get_src(&self, src: NodeId<SrcNodeId>) -> Option<HashSet<NodeId<DstNodeId>>> {
+        self.src.borrow().get(&src).cloned()
+    }
+
+    /// Return a set of all nodes mapped to `src`.
+    pub fn get_dst(&self, dst: NodeId<DstNodeId>) -> Option<HashSet<NodeId<SrcNodeId>>> {
+        self.dst.borrow().get(&dst).cloned()
+    }
+
+    /// Return a set of all mapped `src` nodes.
+    pub fn get_srcs(&self) -> HashSet<NodeId<SrcNodeId>> {
+        self.src.borrow().keys().into_iter().cloned().collect()
+    }
+
+    /// Return a set of all mapped `dst` nodes.
+    pub fn get_dsts(&self) -> HashSet<NodeId<DstNodeId>> {
+        self.dst.borrow().keys().into_iter().cloned().collect()
+    }
+
+    /// `true` if `src` is mapped to exactly one `dst` node, and vice versa.
+    ///
+    /// Returns `false` if `src` is not mapped at all.
+    pub fn is_src_unique(&self, src: NodeId<SrcNodeId>) -> bool {
+        if !self.contains_src(src) {
+            return false;
+        }
+        (*self.src.borrow().get(&src).unwrap()).len() == 1
+        && self.dst
```

Comment:
The formatting here is a bit weird: I missed that line 146 is a continuation of 145.